### PR TITLE
Use Reference from jujubundlelib as a parameter.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/49/
Test PASSed.

File: theblues/charmstore.py [hunk 1]
```diff
@@ -70,10 +72,14 @@ def _get(self, url):
     def _meta(self, entity_id, includes):
         '''Retrieve metadata about an entity in the charmstore.
 
-        @param entity_id The ID of the entity to get.
+        @param entity_id The ID of the entity to get either a reference
```
Comment (mhilton) [inline theblues/charmstore.py]: either as a ... like below?

File: theblues/charmstore.py [hunk 2]
```diff
@@ -103,11 +109,14 @@ def entity(self, entity_id, get_files=False):
     def entities(self, entity_ids):
         '''Get the default data for entities.
 
-        @param entity_ids A list of entity id strings.
+        @param entity_ids A list of entity id strings or references.
```
Comment (mhilton) [inline theblues/charmstore.py]: A list of entity ids either as a references or strings. ?

File: theblues/charmstore.py [hunk 3]
```diff
@@ -147,12 +159,18 @@ def bundle_visualization(self, bundle_id):
         return response.content
 
     def bundle_visualization_url(self, bundle_id):
-        url = '%s/%s/diagram.svg' % (self.url, bundle_id)
+        if isinstance(bundle_id, references.Reference):
```
Comment (mhilton) [inline theblues/charmstore.py]: this is a quite common pattern, is it worth factoring out into a function?

**Comment by mhilton** [comment]
:+1: with some thoughts (take or leave)

File: theblues/charmstore.py (hunk 1 above)
Comment (fabricematrat) [inline theblues/charmstore.py]: Done.

File: theblues/charmstore.py (hunk 2 above)
Comment (fabricematrat) [inline theblues/charmstore.py]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/50/
Test PASSed.

File: theblues/charmstore.py [hunk 4]
```diff
@@ -70,10 +72,14 @@ def _get(self, url):
     def _meta(self, entity_id, includes):
         '''Retrieve metadata about an entity in the charmstore.
 
-        @param entity_id The ID of the entity to get.
+        @param entity_id The ID either a reference or a string of the entity
+               to get.
         @param includes Which metadata fields to include in the response.
         '''
-        url = '%s/%s/meta/any?' % (self.url, entity_id)
+        if isinstance(entity_id, references.Reference):
```
Comment (urosj) [inline theblues/charmstore.py]: isinstance? Why not is_instance or isInstance (not part of this PR, but still)

File: theblues/charmstore.py (hunk 3 above)
Comment (urosj) [inline theblues/charmstore.py]: +1

**Comment by urosj** [comment]
Refactoring the isinstance part and and obtaining what you need as a result would simplify the code quite a bit.

File: theblues/charmstore.py (hunk 3 above)
Comment (fabricematrat) [inline theblues/charmstore.py]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/51/
Test PASSed.

File: theblues/charmstore.py [hunk 5]
```diff
@@ -375,3 +379,16 @@ def fetch_macaroon(self):
...
+
+
+def _get_path(entity_id):
+    '''Get the entity_id as a string if it is a Reference.
+
+    @param entity_id The ID either a reference or a string of the entity
+          to get.
+    @returns: entity_id as a string
+    '''
+    if isinstance(entity_id, references.Reference):
```
Comment (frankban) [inline theblues/charmstore.py]: Not a review, but just glanced at this, please use isinstance only when really necessary:

``` python
try:
    path = entity_id.path()
except AttributeError:
    path = entity_id
return path
```

Comment (fabricematrat) [inline theblues/charmstore.py]: Good suggestion.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/52/
Test PASSed.

**Comment by fabricematrat** [comment]
:shipit: thanks @mhilton && @urosj  && @frankban

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/theblues-lib-merge
